### PR TITLE
New Module: Snippy

### PR DESCRIPTION
**Pull request by ktmeaton**
This PR adds a module for [Snippy](https://github.com/tseemann/snippy), a rapid haploid variant calling and core genome alignment tool. The module parses the ```txt``` output of the commands ```snippy``` and ```snippy-core```. The parsed data is added to the general stats table, and equivalent plots are produced as report sections.

 - [x] This comment contains a description of changes (with reason)
 - [x] `CHANGELOG.md` has been updated

 - [x] There is example tool output for tools in the https://github.com/ewels/MultiQC_TestData repository or attached to this PR
 - [x] Code is tested and works locally (including with `--lint` flag)
 - [x] `docs/README.md` is updated with link to below
 - [x] `docs/modulename.md` is created
 - [x] Everything that can be represented with a plot instead of a table is a plot
 - [x] Report sections have a description and help text (with `self.add_section`)
 - [x] There aren't any huge tables with > 6 columns (explain reasoning if so)
 - [x] Each table column has a different colour scale to its neighbour, which relates to the data (eg. if high numbers are bad, they're red)
  *I struggle with... color. If you have suggestions or improvements please let me know!*
 - [x] Module does not do any significant computational work

Example Files:

  - [snippy-core1.txt](https://github.com/ewels/MultiQC/files/5340837/snippy-core1.txt)
  - [GCA_009669565.1_ASM966956v1_genomic.txt](https://github.com/ewels/MultiQC/files/5340836/GCA_009669565.1_ASM966956v1_genomic.txt)
  - [GCA_009669545.1_ASM966954v1_genomic.txt](https://github.com/ewels/MultiQC/files/5340838/GCA_009669545.1_ASM966954v1_genomic.txt)
  - [GCA_009669555.1_ASM966955v1_genomic.txt](https://github.com/ewels/MultiQC/files/5340839/GCA_009669555.1_ASM966955v1_genomic.txt)

Example Report:
  - [multiqc_report_snippy.zip](https://github.com/ewels/MultiQC/files/5341800/multiqc_report_snippy.zip)

### CODE REVIEW

**Comment by ktmeaton** [comment]
The step **All modules / Custom report filename** is taking a very long time (1h+) with command:

```multiqc --lint test_data/data/modules/ --filename full_report.html```

I'm guessing it's because my search_patterns are too vague? I'll troubleshoot locally.

**Comment by ewels** [comment]
> The step **All modules / Custom report filename** is taking a very long time (1h+) with command:
> 
> `multiqc --lint test_data/data/modules/ --filename full_report.html`
> 
> I'm guessing it's because my search_patterns are too vague? I'll troubleshoot locally.

I just hit exactly the same issue in another module. In the end I tracked the problem down to having a `contents_re` regex that _starts_ with a `.*`. This means that it has to search through all of every file character by character to be sure that the start of the regex doesn't match, which is super slow. Basically anything other than this at the start of the regex will fail almost immediately on the majority of files so is a lot faster. See https://github.com/ewels/MultiQC/pull/1370#issuecomment-792355305 for other discussion 👈🏻

**Comment by ktmeaton** [comment]
I'm revisiting this PR and attempting to rebase it to master.

**Comment by ktmeaton** [comment]
Sorry I didn't lint these updates, thanks for cleaning up the formatting!

File: multiqc/utils/search_patterns.yaml [hunk 1]
```diff
@@ -632,6 +632,12 @@ slamdunk/tcperreadpos:
 slamdunk/tcperutrpos:
   contents: "# slamdunk tcperutr"
   num_lines: 1
+snippy/snippy:
+  contents: "VariantTotal"
```
Comment (ewels) [inline multiqc/utils/search_patterns.yaml:636]: Any reason not to look for `snippy`?

```
Software       snippy 4.6.0
```

**Comment by ewels** [comment]
Hi @ktmeaton,

Apologies for the very slow review. Getting to it now. Code looks good and module output is fine.

I have one main question / concern: the General Statistics table columns seem to mirror the data in the bar graphs almost exactly. Typically modules should contribute 0-3 columns to General Stats - only the most important numbers that are useful to compare to output from other tools.

I'm tempted to just drop the general stats numbers completely, given that everything is already visible in the plots. Would you be ok with this? Are there any key columns that are so important that you think it's worth repeating them?

Thanks,

Phil

**Comment by ewels** [comment]
Hi @ktmeaton,

I'm keen to get a MultiQC release out in the coming few days so have gone ahead and made the changes I suggested so that I can get this merged and included in version 1.11.

Please have a look over the changes I have made and let me know if there are any problems - feel free to make a follow-up pull-request if anything needs fixing.

Many thanks for writing the new module!

Phil

**Comment by ktmeaton** [comment]
Hi @ewels,

Thanks for cutting down on the redundant headers in the table! I wasn't sure where to make the cuts (tables or figures) in the original PR so I put ALL the statistics in ALL the places :) This looks great, thank you so much!

Kat

**Comment by ewels** [comment]
haha, I think this is a pretty widely relevant meme for many MultiQC pull requests 😆 

image

No worries though - this is often one of my main jobs when reviewing. Trying to stay true to the MultiQC mantra that reports are for oversight only and should only show the minimum useful information 🕵🏻

**Pull request merged**